text/0000-typed-array-conveniences.md

```diff
@@ -0,0 +1,312 @@
+    // Constructs a new typed array for this region.
+    pub fn to_typed_array<'c, C>(self, cx: &mut C) -> JsResult<'c, JsTypedArray<T>>
+    where C: Context<'c>;
+}
+```
+
+## Convenient queries
+
+This RFC also defines a number of methods that query typed arrays for metadata.
+
+### `JsTypedArray`
+
+```rust
+impl<T: Binary> JsTypedArray<T> {
+    // Returns the region of the backing buffer for this typed array.
+    pub fn to_region<'cx, C>(
+        &self,
+        cx: &mut C,
+    ) -> Region<'cx, T>
+    where
+        C: Context<'cx>;
+
+    // Returns the backing buffer.
+    pub fn buffer<'cx, C>(
+        &self,
+        cx: &mut C,
+    ) -> Handle<'cx, JsArrayBuffer>
+    where
+        C: Context<'cx>;
+
+    // Returns the byte offset into the backing buffer.
+    pub fn byte_offset<'cx, C>(
+        &self,
+        cx: &mut C,
+    ) -> usize
+    where
+        C: Context<'cx>;
+
+    // Returns the number of elements in the typed array.
+    pub fn len<'cx, C>(
+        &self,
+        cx: &mut C,
+    ) -> usize
+    where
+        C: Context<'cx>;
+}
+```
+
+### `TypedArray`
+
+This RFC adds a convenience method to the `TypedArray` trait:
+
+```rust
+pub trait TypedArray: Sealed {
+    /* as before... */
+
+    // Returns the number of bytes in this typed array.
+    fn byte_length<'cx, C>(&self, cx: &mut C) -> usize
+    where
+        C: Context<'cx>;
+}
+```
+
+### `Region`
+
+The `Region` struct also defines a number of convenient queries.
+
+```rust
+impl<'cx, T: Binary> Region<'cx, T> {
+
+pub fn buffer(self) -> Handle<'cx, JsArrayBuffer>;
+
+pub fn byte_offset(self) -> usize;
+
+pub fn len(self) -> usize;
+
+pub fn byte_length(self) -> usize;
+}
+```
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There aren't any real drawbacks to exposing this functionality, since it's provided by Node-API. There are, however, a number of tradeoffs in the particulars of the design.
+
+The main drawbacks of this RFC's proposed design are:
+
+## Naming inconsistency: `len()` vs `byte_length()`
+
+This RFC goes with `len()` to match a strong Rust convention, and `byte_length()` to match the typed array naming convention (using snake case, of course). These does lead to an inconsistency between the two. A couple alternatives would be `byte_len()` or `size()`. But then `byte_len()` ends up feeling a bit inconsistent with `byte_offset()`. We could then choose `byte_off()` but that's a bit opaque.
+
+## Unfortunate docs placement of `region()`
+
+In order to make the lifetimes work out, `region()` is a method of `Handle<JsArrayBuffer>` instead of `JsArrayBuffer`. This has the unfortunate effect of placing the method in the API docs for [`neon::handle::Handle`][neon::handle::Handle] instead of [`neon::types::JsArrayBuffer`][neon::types::JsArrayBuffer], which makes them less discoverable. The other docs should make an effort to mention the method prominently, use it in examples, and link to its docs in order to mitigate this issue.
```

Comment:
I don't intuitively understand the issue with lifetimes since in `&'a Handle<'cx, Uint8Array>` an `'a` could be chosen that lives as long as `'cx`.

Reply:
This is to allow downcasting *and* returning a region in the same method. Otherwise, you would have a compiler error returning a reference to a dropped value.